### PR TITLE
TEL comments on day 4 homework

### PR DESCRIPTION
**Pull request by tomlatham**
Generally very well done.  There are a few issues here and there, for which you can see my changes and the line-by-line comments.

### CODE REVIEW

File: MPAGSCipher/CipherType.hpp [hunk 1]
```diff
@@ -8,11 +8,11 @@
...
 /**
  * \enum CipherType
- * \brief Enumeration to define the types in which the ciphers can operate
+ * \brief Enumeration to define the types of ciphers that can be used
  */
 enum class CipherType {
-  caesar, ///< Apply caesar cipher
-  playfair  ///< Apply playfair cipher
+  Caesar, ///< Apply caesar cipher
+  Playfair  ///< Apply playfair cipher
```
Comment (tomlatham) [inline MPAGSCipher/CipherType.hpp]: Generally both the enumeration types and each of the possible values are written in UpperCamelCase

File: MPAGSCipher/PlayfairCipher.cpp [hunk 2]
```diff
@@ -6,18 +6,19 @@
 #include<utility>
 #include<map>
 
-PlayfairCipher::PlayfairCipher(const std::string& key): key_{key}
+PlayfairCipher::PlayfairCipher(const std::string& key)
 {
-	setKey(key_);
+	setKey(key);
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: Just to reduce ambiguity about what's happening.

File: MPAGSCipher/PlayfairCipher.cpp [hunk 3]
```diff
@@ -6,18 +6,19 @@
...
 void PlayfairCipher::setKey(const std::string& key)
 {
 	//store the original key
 	key_ = key;
+
 	// Append the alphabet
-	std::string alphabetStr_{alphabet_.begin(),alphabet_.end()};
-	key_.append(alphabetStr_);
+	key_.append("ABCDEFGHIJKLMNOPQRSTUVWXYZ");
+
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: What you did is not wrong but this is perhaps slightly clearer.
See our solution for another approach again, which moves all instances of the alphabet in the whole package into one place.

File: MPAGSCipher/PlayfairCipher.cpp [hunk 4]
```diff
@@ -33,19 +34,16 @@ void PlayfairCipher::setKey(const std::string& key)
 	std::string alreadySeen{""};
 	auto iterDuplicates = std::remove_if(key_.begin(), key_.end(), [&] (char in) {if(alreadySeen.find(in)!=std::string::npos){return true;}else{alreadySeen+=in; return false;}});
 	key_.erase(iterDuplicates, key_.end());
-	// Store the coords of each letter
-
-	std::map<char, std::pair<size_t, size_t>> mapLetter2Coords;
-        std::map<std::pair<size_t, size_t>, char> mapCoords2Letter;
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: You already have private data members, so just use those directly

File: MPAGSCipher/PlayfairCipher.cpp [hunk 5]
```diff
@@ -33,19 +34,16 @@ void PlayfairCipher::setKey(const std::string& key)
...
 
+	// Store the coords of each letter
 	size_t x{0},y{0};
-	std::pair<size_t,size_t> elem{x,y};
+	PlayfairCoords elem{x,y};
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: Use the type definition - see my comments on the header file.

File: MPAGSCipher/PlayfairCipher.cpp [hunk 6]
```diff
@@ -59,22 +57,20 @@ void PlayfairCipher::setKey(const std::string& key)
...
-
-	// Store the playfair cipher key map
-	
-	mapLetter2Coords_ = mapLetter2Coords;
-	mapCoords2Letter_ = mapCoords2Letter;
-		
 }
 
-const std::string PlayfairCipher::applyCipher(const std::string& inputText, const CipherMode cipherMode)
+std::string PlayfairCipher::applyCipher(const std::string& inputText, const CipherMode cipherMode) const
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: Function signature does not match that requested.

File: MPAGSCipher/PlayfairCipher.cpp [hunk 7]
```diff
@@ -59,22 +57,20 @@ void PlayfairCipher::setKey(const std::string& key)
...
-	std::pair<size_t,size_t> firstCoord, secondCoord;
-	size_t store;
-	// Change J->I
+	PlayfairCoords firstCoord, secondCoord;
+	// TEL - there's a lot of duplicated code here due to you introducing the switch this early
+	//     - also, having the replacement of J with I and insertion of X or Q to separate
+	//       identical letters in a digraph only in the encrypt raises some potential problems,
+	//       since you cannot guarantee that the text provided by the user for decryption
+	//       necessarily obeys the rules
+	//     - see our solution for another approach
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: As I say in the comment, the preprocessing should also be applied when decrypting. Of course if it has been properly encrypted by this method then nothing will happen. But the user is at liberty to provide any text that they choose and having a J in the text, for example, will cause serious problems in the execution.

File: MPAGSCipher/PlayfairCipher.cpp [hunk 8]
```diff
@@ -85,42 +81,49 @@ const std::string PlayfairCipher::applyCipher(const std::string& inputText, cons
...
+						input.insert(i+1, x);
 					}
-					else if(input[i] == input[i-1] && input[i] == 'X')
+					else if(input[i] == input[i+1] && input[i] == 'X')
 					{
-						input.insert(i, q);
+						// If this character is the same as
+						// the next and both are X,
+						// insert a Q between them
+						input.insert(i+1, q);
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: This did not quite apply the rule correctly, for example the input BAA would be processed to BAXA, which is not necessary since the two A's are not in the same digraph.

File: MPAGSCipher/PlayfairCipher.cpp [hunk 9]
```diff
@@ -85,42 +81,45 @@ const std::string PlayfairCipher::applyCipher(const std::string& inputText, cons
...
 				}
 
 				// if the size of input is odd, add a trailing Z
-
+				// unless the last character is already a Z,
+				// in which case add an X
 				if(input.length()%2 != 0)
 				{
-					input+="Z";
+					input += (input[input.length()-1] == 'Z') ? 'X' : 'Z';
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: If the last letter is already a Z you need to add something other than a Z or you end up with two identical letters in the same digraph.

File: MPAGSCipher/PlayfairCipher.cpp [hunk 10]
```diff
@@ -130,8 +129,14 @@ const std::string PlayfairCipher::applyCipher(const std::string& inputText, cons
 						firstCoord.first = (firstCoord.first + 1)%5;
 						secondCoord.first = (secondCoord.first + 1)%5;
 					}
-					// 	- Find the letter associated with the new coords
+					else
+					{
+						size_t store = firstCoord.first;
+						firstCoord.first = secondCoord.first;
+						secondCoord.first = store;
+					}
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: Your first test is simply the opposite of the other two combined, so you can make it a simple else.

File: MPAGSCipher/PlayfairCipher.cpp [hunk 11]
```diff
@@ -148,53 +153,33 @@ const std::string PlayfairCipher::applyCipher(const std::string& inputText, cons
...
+						secondCoord.first = (5 + secondCoord.first - 1)%5;
 					}
-					else if(firstCoord.second == secondCoord.second)
+					else
 					{
-						firstCoord.first = (firstCoord.first - 1)%5;
-						secondCoord.first = (secondCoord.first - 1)%5;
+						size_t store = firstCoord.first;
+						firstCoord.first = secondCoord.first;
+						secondCoord.first = store;
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: As per the encrypt you can make the rectangle option a simple else.
One other change is that you need to add 5 before subtracting the 1 (so I suppose you could simply add 4 instead) in order to avoid overflows - it leads to the wrong result.

File: MPAGSCipher/PlayfairCipher.cpp [hunk 12]
```diff
@@ -148,53 +153,33 @@ const std::string PlayfairCipher::applyCipher(const std::string& inputText, cons
...
-					else if(input[i-2] == input[i] && input[i-1] == 'Q')
-					{
-						input.erase(i-1, 1);
-					}
-
-					if(input[input.length()-1] == 'Z')
-					{
-						input.erase(input.length()-1, 1);
-					}
-				}
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: This is beyond the scope of the cipher.  There is no reliable way to do what you're doing since you will encounter cases where you actually remove letters from the input text, e.g. TAX ACT would become TA ACT.

File: MPAGSCipher/PlayfairCipher.cpp [hunk 13]
```diff
@@ -148,53 +153,33 @@ const std::string PlayfairCipher::applyCipher(const std::string& inputText, cons
...
 			}break;
 	}
 	
 	// return the text
 	std::cout << "Playfair cipher applied!" << std::endl;
-	const std::string output = input;
-	return output;
+	return input;
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: Not quite sure why you were doing this

File: MPAGSCipher/PlayfairCipher.hpp [hunk 14]
```diff
@@ -42,22 +42,19 @@ class PlayfairCipher {
      * \param cipherMode whether to encrypt or decrypt the input text
      * \return the result of applying the cipher to the input text
      */
-    const std::string applyCipher( const std::string& inputText, const CipherMode cipherMode );
+    std::string applyCipher( const std::string& inputText, const CipherMode cipherMode ) const;
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.hpp]: As mentioned above, the signature didn't match what was requested.

File: MPAGSCipher/PlayfairCipher.hpp [hunk 15]
```diff
@@ -42,22 +42,19 @@ class PlayfairCipher {
...
   private:
-    /// The alphabet - used to determine the cipher character given the plain character and the key
-    const std::vector<char> alphabet_ = {'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H','I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z'};
-
-    /// The size of the alphabet
-    const std::vector<char>::size_type alphabetSize_ = alphabet_.size();
-
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.hpp]: Given the above change these are no longer necessary

File: MPAGSCipher/PlayfairCipher.hpp [hunk 16]
```diff
@@ -42,22 +42,19 @@ class PlayfairCipher {
...
     /// The cipher key, essentially a constant shift to be applied
     std::string key_ = "A";
 
+    /// Type definition for the coordinates in the 5x5 table
+    using PlayfairCoords = std::pair<size_t,size_t>;
+
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.hpp]: Use a typedef to make the code clearer

File: README.md [hunk 17]
```diff
@@ -101,15 +109,30 @@ MPAGS-Code
...
-├── mpags-cipher.cpp
+├── Testing               Subdirectory for testing the MPAGSCipher library
+│   ├── CMakeLists.txt
+│   ├── catch.hpp
+│   ├── testCaesarCipher.cpp
+│   ├── testCatch.cpp
+│   ├── testHello.cpp
+│   ├── testPlayfairCipher.cpp
+│   ├── testProcessCommandLine.cpp
+│   └── testTransformChar.cpp
```
Comment (tomlatham) [inline README.md]: Ideally the documentation should be updated to reflect the changes

File: Testing/testPlayfairCipher.cpp [hunk 18]
```diff
@@ -0,0 +1,15 @@
...
+
+TEST_CASE("Playfair Cipher encryption", "[playfair]") {
+  PlayfairCipher cc{"hello"};
+  REQUIRE( cc.applyCipher("BOBISSOMESORTOFJUNIORCOMPLEXXENOPHONEONEZEROTHING", CipherMode::Encrypt) == "FHIQXLTLKLTLSUFNPQPKETFENIOLVSWLTFIAFTLAKOWATEQOKPPA");
+}
+
+TEST_CASE("Playfair Cipher decryption", "[playfair]") {
+  PlayfairCipher cc{"hello"};
+  REQUIRE( cc.applyCipher("FHIQXLTLKLTLSUFNPQPKETFENIOLVSWLTFIAFTLAKOWATEQOKPPA", CipherMode::Decrypt) == "BOBISXSOMESORTOFIUNIORCOMPLEXQXENOPHONEONEZEROTHINGZ");
+}
```
Comment (tomlatham) [inline Testing/testPlayfairCipher.cpp]: Introduce some tests for the Playfair cipher. Ideally, these should test all the corner cases (the ones I've added here (copied from our solution) do test some but not all).

File: mpags-cipher.cpp [hunk 19]
```diff
@@ -18,7 +18,7 @@ int main(int argc, char* argv[])
   const std::vector<std::string> cmdLineArgs {argv, argv+argc};
 
   // Options that might be set by the command-line arguments
-  ProgramSettings settings { false, false, "", "", "", CipherMode::Decrypt, CipherType::caesar};
+  ProgramSettings settings { false, false, "", "", "", CipherMode::Encrypt, CipherType::Caesar};
```
Comment (tomlatham) [inline mpags-cipher.cpp]: The documentation and help message say that encryption is the default but it was actually decryption - a bit confusing, I initially thought that it wasn't working at all!

File: mpags-cipher.cpp [hunk 20]
```diff
@@ -133,13 +133,13 @@ int main(int argc, char* argv[])
 
   switch(settings.cipherType)
   {
-	case CipherType::caesar:
+	case CipherType::Caesar:
 		{
-			outputText = caesarcipher.CaesarCipher::applyCipher(inputText, settings.cipherMode);
+			outputText = caesarcipher.applyCipher(inputText, settings.cipherMode);
```
Comment (tomlatham) [inline mpags-cipher.cpp]: When calling a member function you don't need to do this scope resolution.
